### PR TITLE
✅ Add remote state upgrader

### PR DESCRIPTION
**Pull request by deifyed**
Adds an upgrade for moving an okctl environment to a remote state situation. See readme for more details

Do not merge until remote state is complete

### CODE REVIEW

File: gists/remote-state-upgrade/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+your machine and run `tar xvf remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz`
+
+Run the following and verify that the checksum matches `52d9434c30638f47b482326b3ea8319758dc7a2cabc5930954334bee776f88c1`
+```shell
+sha256sum remote-state-upgrader
+```
+
+
+### Linux 
+Download [remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz](dist/remote-state-upgrader_v0.0.0_Linux_amd64.tar.gz) to
```
Comment (yngvark) [inline gists/remote-state-upgrade/README.md:20]: darwin -> linux

File: gists/remote-state-upgrade/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+```shell
+sha256sum remote-state-upgrader
+```
+
+## Usage
+
+1. Authenticate with your desired okctl environment by either 
+   1. Running `okctl venv -c <cluster manifest>` (preferred)
+   2. Running `saml2aws login`, see [saml2aws](https://github.com/Versent/saml2aws)
+2. Run `remote-state-upgrade <path to cluster manifest>`
```
Comment (yngvark) [inline gists/remote-state-upgrade/README.md:33]:
```suggestion
2. Run `./remote-state-upgrade <path to cluster manifest>`
```

for those not accustomed to CLI.

File: gists/remote-state-upgrade/main.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+Usage: ./remote-state-upgrade <path to cluster manifest>
+Example: ./remote-state-upgrade cluster.yaml
+`
+
+type upgradeInfoMessageOpts struct {
+	ClusterName string
+}
+
+const upgradeInfoMessage = `
```
Comment (yngvark) [inline gists/remote-state-upgrade/main.go:101]: kudos for ux

File: gists/remote-state-upgrade/main.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"text/template"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+)
+
+func main() {
```
Comment (yngvark) [inline gists/remote-state-upgrade/main.go:14]: A bit long function, could be written for instance like this

```go

func main() {
	if len(os.Args) != 2 {
		fmt.Print(usage)

		os.Exit(1)
	}

	preflightCheck()

	log.Println("Parsing cluster manifest")

	clusterManifest, err := parseClusterManifest(os.Args[1])
	if err != nil {
		if errors.Is(err, ErrNotFound) {
			log.Fatal(t(ClusterManifestNotFound))
		}

		log.Fatal(fmt.Errorf("acquiring cluster name: %w", err))
	}

	err = printInfoMsg(err, clusterManifest) // The idea is to group similar things to make the calling function (main) simpler to read
	if err != nil {
		log.Fatal(fmt.Errorf("printing info msg: %w", err))
	}

	doUpgrade(clusterManifest) // Same idea
}
```

**Review by yngvark — APPROVED** [review]
Straight forward. Looks good!

File: gists/remote-state-upgrade/README.md (hunk 1 above)
Comment (deifyed) [reply gists/remote-state-upgrade/README.md:20]: Nice catch, thanks

File: gists/remote-state-upgrade/README.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+## Installation
+
+### MacOs
+Download [remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz](dist/remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz) to
+your machine and run 
+```shell
+tar xvf remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz
+```
+
+Run the following and verify that the checksum matches
```
Comment (kmoberg) [inline gists/remote-state-upgrade/README.md:16]: "Run the following" what? 
Should probably add `md5 remote-state-upgrader_v.0.0.0_Darwin_amd64.tar.gz`

File: gists/remote-state-upgrade/README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+### Linux 
+Download [remote-state-upgrader_v0.0.0_Linux_amd64.tar.gz](dist/remote-state-upgrader_v0.0.0_Linux_amd64.tar.gz) to
+your machine and run 
+
+```shell
+tar xvf remote-state-upgrader_v0.0.0_Darwin_amd64.tar.gz
+```
+
+Run the following and verify that the checksum matches
```
Comment (kmoberg) [inline gists/remote-state-upgrade/README.md:32]: Same here.

File: gists/remote-state-upgrade/errors.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	ClusterManifestNotFound
+	NotAuthenticated
+)
+
+func t(key int) string {
+	switch key {
+	case ClusterManifestNotFound:
+		return "Cluster manifest not found at provided path, are you sure it is correct?"
+	case BucketAlreadyExists:
+		return "The bucket already exists. To not overwrite any important data, please contact #kjøremiljø"
```
Comment (kmoberg) [inline gists/remote-state-upgrade/errors.go:27]: Clarifying language a bit, should probably say:
"To avoid overwriting any important data"

**Review by kmoberg — APPROVED** [review]
Nice! A couple of minor language changes could make it more DX/UX friendly.

**Review by kmoberg — APPROVED** [review]

**Comment by deifyed** [comment]
Closing this PR without merging due to the realization that I've accidentally implemented this upgrade as the command `okctl maintenance state-upload <path>`.

**Pull request closed without merging**